File: 백준/Diamond/18185. 라면 사기 （Small）/라면 사기 （Small）.cc
#include <iostream>
#include <algorithm>
using namespace std;
int N;
int arr[10003];


int tri(int i){
    int cnt = min(arr[i], min(arr[i+1], arr[i+2]));
    for (int j = 0; j < 3; ++j) {
        arr[i + j] -= cnt;
    }
    return cnt * 7;
}

int bi(int i, int cnt){
    for (int j = 0; j < 2; ++j) {
        arr[i + j] -= cnt;
    }
    return cnt * 5;
}

int solve(){
    int total = 0;
    for (int i = 0; i <= N; ++i){
        if (arr[i+1] > arr[i+2]){
            int cnt = min(arr[i], arr[i+1] - arr[i+2]);
            total += bi(i,cnt);
            total += tri(i);
        }
        else{
            total += tri(i);
            int cnt = min(arr[i], arr[i+1]);
            total += bi(i,cnt);
        }
        total += arr[i] * 3;
        arr[i] = 0;
    }
    return total;
}

int main(){
    cin >> N;
    for (int i = 0; i < N; ++i) {
        cin >> arr[i];
    }
    cout << solve() << '\n';
}

File: 백준/Diamond/18185. 라면 사기 （Small）/라면 사기 （Small）.py
import sys;input = sys.stdin.readline

def tri(i):
    tmp = min(arr[i:i + 3])
    for j in range(3): arr[i+j] -= tmp
    return tmp * 7

def bi(i):
    tmp = min(arr[i], arr[i + 1])
    for j in range(2): arr[i+j] -= tmp
    return tmp*5

N = int(input())
arr = list(map(int, input().split())) + [0, 0]
total = 0
for i in range(N):
    if arr[i + 1] > arr[i+2]:
        tmp = min(arr[i], arr[i+1] - arr[i+2])
        for j in range(2): arr[i+j] -= tmp
        total += tmp*5
        total += tri(i)
    else:
        total += tri(i)
        total += bi(i)
    total += arr[i] * 3
    arr[i] = 0
print(total)